src/core/ucc_context.c

```diff
@@ -759,9 +779,22 @@ ucc_status_t ucc_context_create(ucc_lib_h lib,
         if (tl_lib->iface->context.create_epilog) {
             status = tl_lib->iface->context.create_epilog(&tl_ctx->super);
             if (UCC_OK != status) {
-                ucc_error("ctx create epilog for %s failed: %s",
-                          tl_lib->iface->super.name, ucc_status_string(status));
-                goto error_ctx_create;
+                if (ucc_tl_is_required(lib, tl_lib->iface, 1)) {
+                    ucc_error("ctx create epilog for %s failed: %s",
+                              tl_lib->iface->super.name, ucc_status_string(status));
+                    goto error_ctx_create;
+                } else {
+                    ucc_debug("ctx create epilog for %s failed: %s",
+                              tl_lib->iface->super.name, ucc_status_string(status));
+                    tl_lib->iface->context.destroy(&tl_ctx->super);
+                    for (j = 0; j < ctx->n_cl_ctx; j++) {
+                        remove_tl_ctx_from_array(ctx->cl_ctx[j]->tl_ctxs,
+                                                 &ctx->cl_ctx[j]->n_tl_ctxs,
+                                                 tl_ctx);
+                    }
+                    remove_tl_ctx_from_array(ctx->tl_ctx, &ctx->n_tl_ctx,
```

Comment:
is tl_ctx is the union of all the cl_ctx[] in this context?

Reply:
it is ALL TL contexts opened at CORE level. In case of failure in epiloge for some tl_ctx i firstly loop through ALL CLs and remote this TL ctx from their tl arrays, and then remove it also from CORE level tl storage